### ISSUE TITLE
False positive in PHPStan 1.4.8 with `assertNotSame` with instances of the same class

### ISSUE THREAD

**Issue by oliverklee** [phpstan/phpstan-phpunit#120]
Using phpstan/phpstan-phpunit 1.0.0 and phpstan/phpstan 1.4.8, the following code creates a false positive:

```php
namespace OliverKlee\Oelib\Tests\Unit\Templating;

use Nimut\TestingFramework\TestCase\UnitTestCase;
use OliverKlee\Oelib\Templating\Template;
use OliverKlee\Oelib\Templating\TemplateRegistry;

class TemplateRegistryTest extends UnitTestCase
{
    /**
     * @test
     */
    public function getForExistingTemplateFileNameCalledTwoTimesReturnsNewInstance(): void
    {
        self::assertNotSame(
            TemplateRegistry::get('EXT:oelib/Tests/Functional/Fixtures/Template.html'),
            TemplateRegistry::get('EXT:oelib/Tests/Functional/Fixtures/Template.html')
        );
    }
}
```

This is the error (with different line numbers):

```
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Tests/Unit/Templating/TemplateRegistryTest.php                                                                                                           
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------- 
  69     Call to static method PHPUnit\Framework\Assert::assertNotSame() with OliverKlee\Oelib\Templating\Template and OliverKlee\Oelib\Templating\Template will  
         always evaluate to false.                                                                                                                                
  91     Call to static method PHPUnit\Framework\Assert::assertNotSame() with OliverKlee\Oelib\Templating\Template and OliverKlee\Oelib\Templating\Template will  
         always evaluate to false.                                                                                                                                
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------- 
```

This problem is new with PHPStan 1.4.8 (as it did no occur with PHPStan 1.4.7).

You can find the complete build with the problem at https://github.com/oliverklee/ext-oelib/pull/904.

(I was not able to use the PHPStan playground for this as I currently do not know of any way of including PHPUnit in it: https://phpstan.org/r/c7fa7bda-1024-45c3-957d-7b91de41b844)

**Comment by herndlm** [comment]
🤦‍♂️ weird, I thought this should not happen because a very similar looking comparison in https://github.com/phpstan/phpstan-src/blob/1.4.8/tests/PHPStan/Rules/Comparison/data/impossible-method-call.php#L75 is fine

> /cc @herndlm Probably worth changing the code to only work for Variable node?
sounds like a solution

**Comment by herndlm** [comment]
Oh look at that, apparently only those static factory-like methods are problems. See https://github.com/phpstan/phpstan-webmozart-assert/pull/122. I'm not sure anymore if limiting it to Variable nodes is the right solution. Maybe there's a bug hidden somewhere else.
I can continue debugging this later today.

**Comment by ondrejmirtes** [comment]
It's not a problem for `New_` thanks to: https://github.com/phpstan/phpstan-src/blob/156cb69be747b7821275803717ef840fae636602/src/Analyser/TypeSpecifier.php#L976-L978

But it's a problem for anything that can be remembered in SpecifiedTypes by TypeSpecifier.

**Comment by herndlm** [comment]
Right, while I was doing a quick test I was actually comparing the type with a node class and thought something is weird. Limiting it to Variable nodes seems to be doing the trick. Give me a couple of minutes..

**Comment by oliverklee** [comment]
@herndlm Thanks for working on this! :heart:

**Comment by herndlm** [comment]
Looks like the same problem existend in a truthy scope with === already. I'm trying to fix that too, and I'm having a hard time creating failing tests for the falsey scope. Which is why this will take a bit longer than a couple of minutes :)

**Comment by herndlm** [comment]
@ondrejmirtes do we need a regression test here?

**Comment by ondrejmirtes** [comment]
Yes, please send it 😊

**Comment by ondrejmirtes** [comment]
Fixed in PHPStan 1.4.9 https://github.com/phpstan/phpstan/releases/tag/1.4.9

**Comment by oliverklee** [comment]
Thanks, @ondrejmirtes and @herndlm! :heart:
